### PR TITLE
New Dataset Sampling Backend: `astartes`

### LINKED ISSUE #426: [BUG]: lost atom mapping information in reaction mode
### Describe the bug
Reaction mode in v1.6 learns slowly or almost does not learn because it drops atom-mapping information.

### Example
I run `chemprop` on the fully atom-mapped GDB reaction set (attached in the comment section) using this script:

<details><summary>cgr.py</summary>
<p>

```python
#!/usr/bin/env python3

import chemprop

arguments = [
    "--data_path", "ccsdtf12_dz.csv",
    "--dataset_type",  "regression",
    "--target_columns", "dE0",
    "--smiles_columns", "rxn_smiles",
    "--metric", "mae",
    "--dropout", "0.05",
    "--epochs", "16",
    "--reaction",
    "--num_folds",  "1",
    "--batch_size", "50",
    "--save_dir", "./"]

args = chemprop.args.TrainArgs().parse_args(arguments)
mean_score, std_score = chemprop.train.cross_validate(args=args, train_func=chemprop.train.run_training)
```

</p>
</details> 

The validation error shows that the learning is slow:

<details><summary>chemprop-1.6 validation mae</summary>
<p>

```
(chemprop-1.6) > grep Val verbose.log 
Validation mae = 15.448567
Validation mae = 14.247193
Validation mae = 13.703250
Validation mae = 13.049585
Validation mae = 12.490423
Validation mae = 12.295789
Validation mae = 11.840348
Validation mae = 11.350094
Validation mae = 11.260314
Validation mae = 11.178272
Validation mae = 10.874099
Validation mae = 10.760612
Validation mae = 10.758510
Validation mae = 11.048934
Validation mae = 10.773161
Validation mae = 10.490051
```

</p>
</details> 

### Expected behavior
With v1.5.0 from pip the results are different and the learning is much faster:

<details><summary>chemprop-1.5 validation mae</summary>
<p>

```
(chemprop-1.5) > grep Val verbose.log 
Validation mae = 13.402517
Validation mae = 10.929379
Validation mae = 9.936616
Validation mae = 8.827146
Validation mae = 8.457737
Validation mae = 7.721934
Validation mae = 7.561652
Validation mae = 7.201899
Validation mae = 6.940796
Validation mae = 6.689141
Validation mae = 6.660027
Validation mae = 6.496381
Validation mae = 6.430068
Validation mae = 6.276500
Validation mae = 6.287375
Validation mae = 6.105415
```

</p>
</details> 

### Possible reason
Here
https://github.com/chemprop/chemprop/blob/3a5782cf791d7b68d9b39f8ed0a7a38235003ff5/chemprop/features/featurization.py#L134-L138
`is_keeping_atom_map()` always returns `False` for the reaction mode. 
Because of this, here
https://github.com/chemprop/chemprop/blob/3a5782cf791d7b68d9b39f8ed0a7a38235003ff5/chemprop/rdkit.py#L27-L29
the atom mapping information is removed.
Finally, this function https://github.com/chemprop/chemprop/blob/3a5782cf791d7b68d9b39f8ed0a7a38235003ff5/chemprop/features/featurization.py#L268 
returns an empty `reac_id_to_prod_id` dict.

### Example
If one adds
```python
print(f'{reac_id_to_prod_id=}\n{only_prod_ids=}\n{only_reac_ids=}\n')
```
above this line https://github.com/chemprop/chemprop/blob/3a5782cf791d7b68d9b39f8ed0a7a38235003ff5/chemprop/features/featurization.py#L298
it prints
```
reac_id_to_prod_id={}
only_prod_ids=[0, 1, 2, 3, 4, 5, 6]
only_reac_ids=[0, 1, 2, 3, 4, 5, 6]
reac_id_to_prod_id={}
only_prod_ids=[0, 1, 2, 3, 4, 5, 6, 7, 8]
only_reac_ids=[0, 1, 2, 3, 4, 5, 6]
```
etc.
Adding the `--keeping_atom_map` argument does not change anything.

### Expected behavior
Modified in the same fashion v1.5 prints
```
reac_id_to_prod_id={0: 2, 1: 3, 2: 4, 3: 5, 4: 6, 5: 7, 6: 8}
only_prod_ids=[0, 1]
only_reac_ids=[]

reac_id_to_prod_id={0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5}
only_prod_ids=[]
only_reac_ids=[]
```
etc.

### Environment
- python version 3.8
- package versions: 
<details><summary>conda list</summary>
<p>

```
# packages in environment at /home/xe/SOFT/miniconda3/envs/chemprop-1.6:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
alabaster                 0.7.13                   pypi_0    pypi
babel                     2.12.1                   pypi_0    pypi
blinker                   1.6.2                    pypi_0    pypi
ca-certificates           2023.05.30           h06a4308_0  
certifi                   2023.7.22                pypi_0    pypi
charset-normalizer        3.2.0                    pypi_0    pypi
chemprop                  1.6.0                    pypi_0    pypi
click                     8.1.6                    pypi_0    pypi
cloudpickle               2.2.1                    pypi_0    pypi
cmake                     3.27.0                   pypi_0    pypi
contourpy                 1.1.0                    pypi_0    pypi
cycler                    0.11.0                   pypi_0    pypi
descriptastorus           2.6.0                    pypi_0    pypi
docstring-parser          0.15                     pypi_0    pypi
docutils                  0.20.1                   pypi_0    pypi
exceptiongroup            1.0.4            py38h06a4308_0  
filelock                  3.12.2                   pypi_0    pypi
flask                     2.3.2                    pypi_0    pypi
fonttools                 4.41.1                   pypi_0    pypi
future                    0.18.3                   pypi_0    pypi
hyperopt                  0.2.7                    pypi_0    pypi
idna                      3.4                      pypi_0    pypi
imagesize                 1.4.1                    pypi_0    pypi
importlib-metadata        6.8.0                    pypi_0    pypi
importlib-resources       6.0.0                    pypi_0    pypi
iniconfig                 1.1.1              pyhd3eb1b0_0  
itsdangerous              2.1.2                    pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
joblib                    1.3.1                    pypi_0    pypi
kiwisolver                1.4.4                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1  
libffi                    3.4.4                h6a678d5_0  
libgcc-ng                 11.2.0               h1234567_1  
libgomp                   11.2.0               h1234567_1  
libstdcxx-ng              11.2.0               h1234567_1  
lit                       16.0.6                   pypi_0    pypi
markupsafe                2.1.3                    pypi_0    pypi
matplotlib                3.7.2                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
mypy-extensions           1.0.0                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0  
networkx                  3.1                      pypi_0    pypi
numpy                     1.24.4                   pypi_0    pypi
nvidia-cublas-cu11        11.10.3.66               pypi_0    pypi
nvidia-cuda-cupti-cu11    11.7.101                 pypi_0    pypi
nvidia-cuda-nvrtc-cu11    11.7.99                  pypi_0    pypi
nvidia-cuda-runtime-cu11  11.7.99                  pypi_0    pypi
nvidia-cudnn-cu11         8.5.0.96                 pypi_0    pypi
nvidia-cufft-cu11         10.9.0.58                pypi_0    pypi
nvidia-curand-cu11        10.2.10.91               pypi_0    pypi
nvidia-cusolver-cu11      11.4.0.1                 pypi_0    pypi
nvidia-cusparse-cu11      11.7.4.91                pypi_0    pypi
nvidia-nccl-cu11          2.14.3                   pypi_0    pypi
nvidia-nvtx-cu11          11.7.91                  pypi_0    pypi
openssl                   3.0.9                h7f8727e_0  
packaging                 23.1                     pypi_0    pypi
pandas                    2.0.3                    pypi_0    pypi
pandas-flavor             0.6.0                    pypi_0    pypi
parameterized             0.8.1              pyhd3eb1b0_1  
pillow                    10.0.0                   pypi_0    pypi
pip                       23.2.1           py38h06a4308_0  
pluggy                    1.0.0            py38h06a4308_1  
protobuf                  4.23.4                   pypi_0    pypi
py4j                      0.10.9.7                 pypi_0    pypi
pygments                  2.15.1                   pypi_0    pypi
pyparsing                 3.0.9                    pypi_0    pypi
pytest                    7.4.0            py38h06a4308_0  
python                    3.8.17               h955ad1f_0  
python-dateutil           2.8.2                    pypi_0    pypi
pytz                      2023.3                   pypi_0    pypi
rdkit                     2023.3.2                 pypi_0    pypi
readline                  8.2                  h5eee18b_0  
requests                  2.31.0                   pypi_0    pypi
scikit-learn              1.3.0                    pypi_0    pypi
scipy                     1.10.1                   pypi_0    pypi
setuptools                68.0.0           py38h06a4308_0  
six                       1.16.0                   pypi_0    pypi
snowballstemmer           2.2.0                    pypi_0    pypi
sphinx                    7.1.1                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.4                    pypi_0    pypi
sphinxcontrib-devhelp     1.0.2                    pypi_0    pypi
sphinxcontrib-htmlhelp    2.0.1                    pypi_0    pypi
sphinxcontrib-jsmath      1.0.1                    pypi_0    pypi
sphinxcontrib-qthelp      1.0.3                    pypi_0    pypi
sphinxcontrib-serializinghtml 1.1.5                    pypi_0    pypi
sqlite                    3.41.2               h5eee18b_0  
sympy                     1.12                     pypi_0    pypi
tensorboardx              2.6.1                    pypi_0    pypi
threadpoolctl             3.2.0                    pypi_0    pypi
tk                        8.6.12               h1ccaba5_0  
tomli                     2.0.1            py38h06a4308_0  
torch                     2.0.1                    pypi_0    pypi
tqdm                      4.65.0                   pypi_0    pypi
triton                    2.0.0                    pypi_0    pypi
typed-argument-parser     1.8.1                    pypi_0    pypi
typing-extensions         4.7.1                    pypi_0    pypi
typing-inspect            0.9.0                    pypi_0    pypi
tzdata                    2023.3                   pypi_0    pypi
urllib3                   2.0.4                    pypi_0    pypi
werkzeug                  2.3.6                    pypi_0    pypi
wheel                     0.38.4           py38h06a4308_0  
xarray                    2023.1.0                 pypi_0    pypi
xz                        5.4.2                h5eee18b_0  
zipp                      3.16.2                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_0  
```

</p>
</details> 

- OS: OpenSUSE 15.4 / Linux 5.14.21

### Checklist
- [x] all dependencies are satisifed: `conda list` shows the packages listed in the `README`
- [x] the unit tests are working: `pytest -v` reports no errors

### PR DESCRIPTION
**Pull request by JacksonBurns**
# TL;DR: New dataset samplers, need reviewer input

## Description
@kspieks and I have released [`astartes`](https://github.com/JacksonBurns/astartes), a package for generating data splits for machine learning. Many of the sampling methods used in `chemprop` have already been implemented over there but also including some more rigorous input validation, user warnings, and an explicit focus on reproducibility.

This PR replaces the __random__ and __scaffold__ samplers with their corresponding implementation in `astartes` and adds two new sampling algorithms to `chemprop`: __KMeans__ and __Kennard-Stone__:
 - As I have added to the README, the KMeans sampler performs K-Means clustering on the morgan fingerprint representation of the molecules in the dataset and then partitions them into training, validation, and testing separately to rigorously evaluate model extrapolation.
 - Kennard-Stone does the opposite, instead uniformly sampling feature space into training and interpolating for validation and testing, which gives and accurate idea of model inference capacity. 
 
The former is great for evaluating models used for discovery, and the latter is better for inference and/or small datasets (with especially interesting applications to the prediction of spectra, see [here](https://nirpyresearch.com/kennard-stone-algorithm/) for a starting point).

## Example
I have added simple unit tests for the new samplers and suggest [this review](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6173947/) for a starting point on why KMeans and Kennard-Stone are worthwhile additions to `chemprop`.

## Question/Reviewer Note
Locally, this change is failing many regression tests. Most of these are because the regression tests were using random sampling, and `astartes` uses the `numpy` random module rather than the Python one resulting in different model performance. Most of these are just outside of the acceptable delta, with some exceptions that will show in the logs. I will note that `astartes` uses the `numpy` recommended approach for random number generation, which we believe is better for reproducibility. I have not updated the 'target' values and instead welcome comments on this in particular.

There were also a couple test failures related to scaffold generation for atom-mapped SMILES (see #426 and linked issues). I have adapted some changes by @shihchengli in #427 to deal with this, and if merged this PR may supercede #427 and resolve #426.

Also not totally sure who to request as reviewer - starting with @kevingreenman and @shihchengli but any who are familiar, please join in.

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by JacksonBurns** [comment]
> Thanks for the PR @JacksonBurns, having more split options available in Chemprop would be great.

Thanks!

> Regarding the test failures, I'm wondering if we should wait a few months and introduce this in chemprop v2 since the change in random number generation will make results inconsistent across different versions of v1.

Ideally I would like to at least get Kennard-Stone and KMeans intro production before v2 so that we can iron out any bugs. Maybe we can merge this PR as-is into v2 and cherry pick the Kennard-Stone and KMeans samplers into the next minor release of v1?

**Comment by davidegraff** [comment]
i'm fine with that strategy- my one ask is that we go all-in for v2 if we decide on this. I.e., either have _all_ of our splitting handled by `astartes` or none of it. It would be nice to absolve `chemprop` of internal data splitting responsibility and make the code more thematically consistent.

**Comment by JacksonBurns** [comment]
I have removed the changes to the random and scaffold sampling backends. I have also made a new branch aimed at merging into v2 that contains the original changes. The `cv` sampler can be replaced with repeat calls to `astartes`'s `train_test_split` and the `random_with_repeated_smiles` can be replaced using KMeans with Levenstein distance on the SMILES strings (and n_clusters=n_unique_smiles_strings) or by implementing the existing sampling approach in `astartes`.

**Comment by JacksonBurns** [comment]
@kevingreenman this should now be ready for review

**Comment by JacksonBurns** [comment]
Closing in favor of #433 rather than maintaining the backported version as well
